cpp/src/lists/copying/gather.cu

```diff
@@ -0,0 +1,195 @@
+  }
+};
+
+/**
+ * @copydoc cudf::lists::detail::gather_list_leaf
+ *
+ */
+std::unique_ptr<column> gather_list_leaf(column_view const& column,
+                                         gather_data const& gd,
+                                         cudaStream_t stream,
+                                         rmm::mr::device_memory_resource* mr)
+{
+  // gather map for our child
+  auto child_gather_map = thrust::make_transform_iterator(
+    thrust::make_counting_iterator<size_type>(0), list_gatherer{gd});
+
+  // size of the child gather map
+  size_type child_gather_map_size;
+  CUDA_TRY(cudaMemcpy(&child_gather_map_size,
+                      gd.offsets + gd.base_offsets->size(),
+                      sizeof(size_type),
+                      cudaMemcpyDeviceToHost));
+
+  // otherwise, just call a regular gather
+  auto leaf_column =
+    cudf::type_dispatcher(column.type(),
+                          cudf::detail::column_gatherer{},
+                          column,
+                          child_gather_map,
+                          child_gather_map + child_gather_map_size,
+                          // note : we don't need to bother checking for out-of-bounds here since
+                          // our inputs at this stage aren't coming from the user.
+                          false,
+                          stream,
+                          mr);
+
+  // the column_gatherer doesn't create the null mask because it expects
+  // that will be done in the gather_bitmask() step.  however, gather_bitmask()
+  // only happens at the root level, and by definition this column is a
+  // leaf.  so we have to generate the bitmask ourselves.
+  // TODO : it might make sense to expose a gather() function that takes a column_view and
+  // returns a column that does this work correctly.
+  rmm::device_buffer null_mask{0, stream, mr};
+  size_type null_count = column.null_count();
+  if (null_count > 0) {
+    auto list_cdv = column_device_view::create(column);
+    auto validity = cudf::detail::valid_if(
+      child_gather_map,
+      child_gather_map + child_gather_map_size,
+      [cdv = *list_cdv] __device__(int index) { return cdv.is_valid(index) ? true : false; },
+      stream,
+      mr);
+
+    leaf_column->set_null_mask(std::move(validity.first), validity.second);
+  }
+
+  return std::move(leaf_column);
+}
+
+/**
+ * @copydoc cudf::lists::detail::gather_list_nested
+ *
+ */
+std::unique_ptr<column> gather_list_nested(cudf::lists_column_view const& list,
+                                           gather_data& parent,
+                                           cudaStream_t stream,
+                                           rmm::mr::device_memory_resource* mr)
+{
+  auto gather_map_begin = thrust::make_transform_iterator(
+    thrust::make_counting_iterator<size_type>(0), list_gatherer{parent});
+
+  // size of the child gather map
+  size_type gather_map_size = 0;
+  CUDA_TRY(cudaMemcpy(&gather_map_size,
```

Comment:
Use `cudaMemcpyAsync` with the passed in `stream` parameter.

Reply:
Whoops.  Good call.